B-Spline - Curvature

christophKaiser opened the pull request:
### What's done until now
Changes in current code of `SplineConverter.h`
- In commit 76041c3, the loading process of B-Spline control points was moved into a separate function.
- In commits 6bbf52f and 248cd62, the calculation of one single curve point was moved into a separate function.
- In commits 3626bff and 95a3ecb, the calculation of the basis functions was spited into smaller parts, which will be needed for the calculation of the derivatives.

New functionality in `SplineConverter.h`
- Commits 83ae485 and 9fd6590 added functions to calculate the first and second derivative of a B-Spline.
- Commits c8ff579 and 8248ab2 add a main function `computeCurvatureOfBSplineCurveWithKnots` which calculates the curvature of a B-Spline. As well, the corresponding curve length is calculated, which is necessary to display or analyse the curvature in dependency to the true curve length. The parameter t along a B-Spline curve with open uniform knot vector doesn't represent the true length along the curve because the calculated curve points aren't evenly spaced. The function `computeCurvatureOfBSplineCurveWithKnots` is independent from an ifc entity because upcoming functions will have a `std::vector<carve::geom::vector<3>>` variable, which will be given to `computeCurvatureOfBSplineCurveWithKnots`.

### Pending work
- [done] The function `computeCurvatureOfIfcBSplineCurveWithKnots` has to be called from somewhere in the project, where the vertices in the `std::vector<carve::geom::vector<3>>` have to be obtained.
- [done] Its result, the curvature, must be interpreted.
- The final result should be displayed on the viewport.

[Update from 2020-08-20 see [below](https://github.com/tumcms/Open-Infra-Platform/pull/202#issuecomment-677907361)]
[Update from 2020-08-30 see [below](https://github.com/tumcms/Open-Infra-Platform/pull/202#issuecomment-683414473)]
[Update from 2020-11-23 see [below](https://github.com/tumcms/Open-Infra-Platform/pull/202#issuecomment-732456798)]

christophKaiser committed:
Added function computePointOfBSpline.
This function will be needed for calculation of the derivatives (coming soon).
At the moment, this change is only the substitution of some calculations.

christophKaiser committed:
Added function obtainBasisFunctionFirstOrder.
This function will be needed for calculation of the derivatives (coming soon).
At the moment, this change is only the substitution of some calculations.

christophKaiser committed:
Added function obtainBasisFunctionNextOrder.
This function will be needed for calculation of the derivatives (coming soon).
At the moment, this change is only the substitution of some calculations.

christophKaiser committed:
Added function computePointOfRationalBSpline.
This function will be needed for calculation of the derivatives (coming soon).
At the moment, this change is only the substitution of some calculations.

christophKaiser committed:
Added doxygen comments to the new functions
As well, a small change in obtainBasisFunctionNextOrder: The parameter numBasisFuncs was replaced by the size()-function.

christophKaiser committed:
Added function loadControlPoints
In a coming commit, this function will be called a second time. It should prevent to initialize the curveConverter entity at multiple places in code.

christophKaiser committed:
Added function computeCurvatureOfIfcBSplineCurveWithKnots

christophKaiser committed:
Added functions to calculate first derivative

christophKaiser committed:
Added functions to calculate second derivative

christophKaiser committed:
Added a junk function to call new functions for compiler
The new function computeCurvatureOfIfcBSplineCurveWithKnots isn't part of the current program flow. This junk function has the purpose to call the function for the compiler and get feedback (error messages) from there.
At this commit, the compiler will fail because the function computeCurvatureOfIfcBSplineCurveWithKnots is missing. (Will be added in one of the next commits.)
This commit with the junk function will be reverted later.

christophKaiser committed:
Changed function computeCurvatureOfBSplineCurveWithKnots to be independent from ifc
(previous, the function was called computeCurvatureOfIFcBSplineCurveWithKnots)
With this change, an ifc-entity isn't necessary to call this function. In a coming enhancement a std::vector<carve::geom::vector<3>> will be the input to OIP and should be given to this function.
In the case that the curvature of an ifc-B-Spline is needed, the properties must be obtained from the IfcBSplineCurveWithKnots entity like in junkfunctionToCallForCompiler.

christophKaiser committed:
Add button to menu
At the moment, the button shows only a message box.

pjanck commented on Core/src/IfcGeometryConverter/SplineConverter.h:
```diff
@@ -143,7 +144,110 @@ namespace OpenInfraPlatform
+					 * \return		Vector with curve length
+					 * \return		Vector with curve curvature
+					 */
+					std::tuple<std::vector<double>, std::vector<double>> computeCurvatureOfBSplineCurveWithKnots(
```
My suggestion would be to rather use `std::vector<std::pair<double, double>>` if you're having pairs with "length along the curve" and its corresponding "curvature at that point". Access the member by `.first` and `.second` .

christophKaiser replied:
Changed in commit 637e6d8

christophKaiser committed:
Changed return of computeCurvatureOfBSplineCurveWithKnots ...
from std::tuple<std::vector<double>, std::vector<double>> to std::vector<std::pair<double, double>> because the length and curvature are pairs of data.
Access the members by .first to get the length and .second to get the curvature.

pjanck commented on Core/src/IfcGeometryConverter/SplineConverter.h:
```diff
@@ -198,9 +197,7 @@ namespace OpenInfraPlatform
-						std::vector<double> length;
-						length.reserve(numCurvePoints);
-						std::vector<double> curvature;
+						std::vector<std::pair<double, double>> curvature;
```
As the naming goes, `lengthsWithCurvatures` would be more descriptive, as there are not only curvatures in the vector.

christophKaiser replied:
Done in commit b7aa29d.

pjanck commented on Core/src/IfcGeometryConverter/SplineConverter.h:
```diff
@@ -211,28 +208,33 @@ namespace OpenInfraPlatform
+								length_i = 0;
 							else
-								length[i] = length[i-1] + (curvePoint - curvePointPrevious).length();
+								length_i = curvature[i-1].first + (curvePoint - curvePointPrevious).length();
```
A more concise `C`-style-like expression would be:

    length_i = (i == 0 ? 0 : curvature[i] ... );

christophKaiser replied:
Done in commit b7aa29d.
Because length_i is reduced to one line / statement, change of commit 9860558 is a suggestion from me; however, I'm not sure, whether it might got to complicated to read and understand.

pjanck replied:
With the addition of the comments, it is transparent what is happening in the lines.

christophKaiser committed:
Renamed the return value and applied immediate-if
(both from pjancks suggestion)

christophKaiser committed:
Moved the calculations inside the pair creation.
This change saves two temporary double variables; however, it might be a bit more complicated to read and understand.

christophKaiser committed:
Added the folder \Rendering\Tools to the library
The existing file Tool.h might be used and an additional class will be added at this place.

christophKaiser committed:
Added class SplineInterpretation
Added new .h and .cpp files of class SplineInterpretation with empty constructor;
Included SplineInterpretation.h in MainWindow.h;
Added the constructor call in MainWindow.cpp (which is executed by the ui button in menu 'Tool').

christophKaiser committed:
Added comments to the variable t
... to make clear what t is, and that t changes its value inside the for-loop.

christophKaiser committed:
Added functions convertSketchToAlignment (not finished) and obtainKnotArrayOpenUniform
The call of convertSketchToAlignment() was added in MainWindow.cpp.
The functions convertSketchToAlignment and obtainKnotArrayOpenUniform were declared in SplineInterpretation.h. As well, included std::vector and carve::geom.
The functions convertSketchToAlignment and obtainKnotArrayOpenUniform were implemented in SplineInterpretation.cpp.
At this stage, all information of a B-Spline is stored in variables and can be observed with a break point at line 45 of convertSketchToAlignment. Further actions will be come after some modification in SplineConverter.

christophKaiser committed:
Merge branch 'SplineInterpretation' into addSplineCurvature

christophKaiser committed:
Introduced new class SplineUtilities in SplineConverter.h
Functions from SplineConverterT will be moved in this new class, where they will be called from the class SplineInterpretation without GeometrySettings, UnitConverter, etc.

christophKaiser committed:
Renamed functions in SplineConverter
computeIfcRationalBSplineCurveWithKnots and computeIfcBSplineCurveWithKnots were renamed because they are independent from an ifc-entity.

christophKaiser committed:
Moved functions from SplineConverter into SplineUtilities
Moved functions: computeCurvatureOfBSplineCurveWithKnots, obtainKnotRange, computePointOfBSpline, computePointOfRationalBSpline, computeBSplineBasisFunctions, obtainBasisFunctionFirstOrder, obtainBasisFunctionNextOrder, computePointOfDerivativeOne, computeDerivativeOneBasisFunctions, obtainBasisFunctionDerivativeFirstOrder, obtainBasisFunctionDerivativeOneNextOrder, computePointOfDerivativeTwo, computeDerivativeTwoBasisFunctions, obtainBasisFunctionDerivativeTwoNextOrder
The class SplineUtilities was moved into a new file.

christophKaiser committed:
Added 'const throw(...)' to SplineInterpretation

christophKaiser committed:
Rearranged #include in SplineInterpretation
(in header file only if include is there necessary)

christophKaiser committed:
Fixed runtime error at vector.reserve()
With
vector.reserve(n);
vector[0] = ..;
a runtime error happened.
The code runs with .resize(..).

christophKaiser committed:
Added call of computeCurvatureOfBSplineCurveWithKnots()
In SplineInterpretation, the class SplineUtilities was included and the call of computeCurvatureOfBSplineCurveWithKnots was added.
The include statements in SplineUtilities were commented out because I got an error at compiling SplineInterpretation. At the moment, SplineUtilities gets its include statements from the classes where its header file is included.
Further investigation to include all really needed files / libraries might be reasonable.

christophKaiser committed:
Added function debugFunction_printCurvatureInConsolWindow to SplineInterpretation
Displays the vector of length with curvature in the console in a syntax which can be pasted into MATLAB.

christophKaiser committed:
Completed doxygen comment of the function obtainKnotArrayOpenUniform

pjanck reviewed (commented):
This is starting to take on a really good shape! Good work!

Suggestion: since #220 , nice functionality for automatic testing is available. Perhaps that can help you test multiple scenarios at once.

christophKaiser committed:
Fixed logic mistake in obtainBasisFunctionDerivativeTwoNextOrder
The explanation was as a comment in the code of computeDerivativeTwoBasisFunctions:
// in C++, k goes from 0 to order-1 = degree; k=0 is done with obtain..FirstOrder()
// in mathematical definition, k goes from 1 to order

Here, the case of [mathematical] k=2 has to be treated.

christophKaiser committed:
Added console outputs for debugging
The output is in the right syntax to be pasted into MATLAB.
(The output is commented out; it's still there in case that it will be needed in future once again.)

christophKaiser committed:
Fixed casting error in computeCurvatureOfBSplineCurveWithKnots
By default, the result of 3/2 was saved as int. Now, the result is saved as double and pow works fine.

christophKaiser committed:
Remove junkfunctionToCallForCompiler from SplineConverter
Now the functions computeCurvatureOfBSplineCurveWithKnots is called from the intended location.
(In more detail, see commit aa71bbe 'Added a junk function to call new functions for compiler')

christophKaiser committed:
Add accuracy to SplineInterpretation
Usually, accuracy is the value of GeometrySettings::getPrecision(). Because GeometrySettings isn't available in this class, the value carve::EPSILON which is used there is assigned.

christophKaiser committed:
Moved computeBSplineCurveWithKnots from SplineConverter into SplineUtilities
As well, computeRationalBSplineCurveWithKnots was moved because it's on the same hierarchy level.
The functions computeBSplineCurveWithKnots is needed in SplineUtilities, where its callable from SplineInterpretation.
The number of curve points (numCurvePoints) became a parameter of computeBSplineCurveWithKnots, because SplineInterpretation might call it with another value. As well, accuracy became a parameter of computeBSplineCurveWithKnots, because SplineConverter gets the value of accuracy from the GeometrySettings instance which isn't available in SplineUtilities.

christophKaiser committed:
Removed numControlPoints as parameter
numControlPoints was removed as parameter which was took through multiple function calls. Now, the value is figured out when needed.

christophKaiser committed:
Added function debugFunction_printVectorOfPointsInConsolWindow
This function is used to get the coordinates of a calculated B-Spline in the console.

christophKaiser committed:
Update to added accuracy
Assignment of the value is moved into the header file because 'carve' is available there sine the last commit.

christophKaiser committed:
Added function lengthInXyPlane in SplineUtilities
The function calculates the length of a carve::geom::vector in xy plane. Its use is in computeCurvatueOfBSplineCurveWithKnots, which result is used for analyses of the horizontal alignment where height differences in z doesn't matter.

christophKaiser commented:
What's happened since last description - only relevant / big changes are listed:
(Summary of the commits and one question in 'Pending Work - Now')

### User Interface
With commit bd7e81d, the new functions can be called via the UI: In the menu 'Tools' is a new option 'Convert Stroke to Alignment' (design name - maybe it will change in future).

### New class `SplineInterpretation`
- Commit 81680eb included the [existing] folder `\Rendering\Tools` into the project, where the new class was created (commit 5cfdb97).
- At this place, the main function convertSketchToAlignment calls the function computeCurvatureOfBSplineCurveWithKnots in SplineUtilities - in SplineConverter the junk function, which called the function, isn't any more needed.
- Tow functions `debugFunction_...` are available to print out some information in the console window. To see all lines, you have to change manually the property Layout > Screen Buffer Size > Height from 25 to something greater than 200. The output can be pasted into MATLAB.

### New class `SplineUtilities`
- With commits 724c5ad and d6e88b7, the new class was introduced in the folder `IfcGeometryConverter`. Some functions from SplineConverter were moved into this new class because the functions have to be called without the converter environment.

### Pending Work - Now
- [answered and done] `SplineInterpretation` should get functions to save and load the information of a `std::vector<carve::geom::vector<3>>` into a file. I prefer a simple syntax inside the save file and write an appropriate reader / writer with the class `fstream` (or `ofstream` and `ifstream`) - reading detail in the documentation is pending. At this point my question: Has to be the syntax a specific one - for example from an external software which generates test data, or the appropriate data type - or am I free in my choose?
- `SplineInterpretation` will get further functions to interpret the curvature.

### Pending Work - In (far) future
- SplineInterpretation doesn't appear in doxygen [not urgent]
- In SplineUtilities: Further investigation to include all really needed files / libraries might be reasonable. (in more detail: commit f13a07d) [not urgent]

[Update from 2020-08-30 see [below](https://github.com/tumcms/Open-Infra-Platform/pull/202#issuecomment-683414473)]

christophKaiser committed:
Merge branch 'development' into addSplineCurvature

christophKaiser committed:
Added functions saveControlPointsIntoFile and loadControlPointsFromFile.
The vector of temporary test data was renamed to make a difference between 'to save' and 'loaded'. The spline calculations are done with the loaded one.
A temporary call of the save and load functions was added inside the main function 'convertSketchToAlignment'. In a pending commit, the head of convertSketchToAlignment (about line 34 to 57) will be changed.
The save and load function work in the directory '[build folde]/testdata/StrokeToAlignment/', which has to be created via build of project 'Copy/CopyOpenInfraPlatformUIResources' in Visual Studio.
Two exception handling are missing; however, in this stage everything works fine if no exception case is triggered.

christophKaiser committed:
fix - Added two const statements into existing code.

christophKaiser commented:
What's happened since last description

### Save and Load were added to `SplineInterpretation`

In commit 62e1341, the functions `saveControlPointsIntoFile` and `loadControlPointsFromFile` were added to `SplineInterpretation`. They work in the directory '[build folde]/testdata/StrokeToAlignment/', which has to be created via build of the project 'Copy/CopyOpenInfraPlatformUIResources' in Visual Studio.

One example file is included in the commit. The file is saved as .sketch, but can be opened with every usual text editor. The information is saved as text in syntax of a MATLAB matrix; appropriate, the matrix can be pasted into MATLAB (e.g. for debug or to visualise).

### Pending work - Now
- `SplineInterpretation` will get further functions to interpret the curvature and to create alignment objects.
- Visualisation of the alignment objects in the viewport.

### Pending Work - In (far) future
- In the code of commit 62e1341, two exception handling are missing; however, in this stage everything works fine if no exception case is triggered. In more detail, search for `// ToDo: exception handling` in the code. [not urgent]

[Update from 2020-11-23 see [below](https://github.com/tumcms/Open-Infra-Platform/pull/202#issuecomment-732456798)]

pjanck reviewed (commented):
This looks very promising. I commend you for the abundance of comments in the code!

christophKaiser committed:
Added function obtainControlPoints
Rearranged 'loadControlPointsFromFile' into the new function. Inside the new function is a MessageBox to choice between sketch and load.
'Sketch' isn't implemented.

christophKaiser committed:
Modified computeCurvatureOfBSplineCurveWithKnots
numCurvePoints was changed to a parameter, instead the old temporally definition in properties. In future, numCurvePoints will be calculated before function call.

christophKaiser committed:
Modified convertSketchToAlignment
Rearranged and new definition of properties, according to first approximation and computation of curvature.
At this stage, the function debugFunction_printCurvatureInConsolWindow prints very nice values of curvature. However, the calculation time of computeCurvatureOfBSplineCurveWithKnots might be a little bit long.

christophKaiser committed:
Fixed function computeDerivativeTwoBasisFuncstions
The vector 'basisFuncsBSpline' isn't needed in the calculation of second derivative; removed the variable. However, the behavior of the code shouldn't change because it wasn't used - maybe a little bit faster.
(The mistake might be caused by C&P.)

christophKaiser committed:
Added functions of moving average with variable window size

christophKaiser committed:
Modified functions of moving average
The function movingAverageVariableWindow gets a new overload. One function handles a std::vector<std::par<double,double>> and calls the other [main] function which handles a std::vector<double>. This split is necessary to handle the variable lengthsWithCurvatures, and curvatureChange as well (comes in next commit).
The variable type of variogrammGetRange was changed appropriate the modifications of movingAverageVariableWindow.

christophKaiser committed:
Added function for change of curvature
The change of curvature (curvatureChange) is calculated with the new function numericDerivative.

pjanck reviewed (commented):
I'm interested to see the results of this code.

christophKaiser committed:
Rearranged debug function for output in console window.
Added one function to print the curvatureChange vector.

christophKaiser committed:
Fixed bugs

christophKaiser committed:
Modified variogrammGetRange: Introduced rangeMax
The visualization of calculated test data showed very high values of range (= window size of average), appropriate the curvature and curvatureChange were like mush. According to similar calculations in MATLAB the maximum value of range is set to 100. More detail in the next comment on github.

christophKaiser commented:
The calculations of curvature and curvature change were tested with the following curve:
![0 curve](https://user-images.githubusercontent.com/64251870/95001633-f4e3c580-05cb-11eb-8212-3df683f4d330.png)
The points are a perfect alignment obtained by CAD.

In MATLAB, I have almost the similar functions to compute the B-Spline, its curvature with change of curvature, and smoothing via moving average & variogram variable. The following two charts were my result, one or two weeks ago:
![1 expected curvature](https://user-images.githubusercontent.com/64251870/95001674-69b6ff80-05cc-11eb-8b9f-ea6f70f34ec2.png)
![2 expected curvatureChange](https://user-images.githubusercontent.com/64251870/95001676-6c195980-05cc-11eb-9335-6ccc3778b313.png)
The first one shows the calculated curvature (blue line) with the smoothed data (orange line). The second one shows the change of curvature obtained by numerical derivative (blue line) and after smoothing (orange line).
In summary, the result was very good at this point.

The implementation in OIP is almost like in MATLAB; thus, I expected the same results. However, I got the following one (visualization with MATLAB, but calculation with OIP):
![3 real curvature](https://user-images.githubusercontent.com/64251870/95001720-0b3e5100-05cd-11eb-9fa3-4b0dbee786d8.png)
![4 real curvatureChange](https://user-images.githubusercontent.com/64251870/95001722-0da0ab00-05cd-11eb-8afe-e1093ab73739.png)
As well, the figures show the curvature and the change of curvature. The calculation is displayed in grey, the result of smoothing in black. In total, this result is very bad because the curvature doesn't show the straight section and the change of curvature doesn't say anything useful.

The reason for this result is a very high value of range in the variogram, which is used as window size of the moving average. The low value in MATLAB might be caused by the peaks to +3.2 and -3.5 near the start and end of the curvature plot; in the B-Spline calculation of OIP, this peak doesn't appear.
The high value of range in OIP isn't caused by a mistake in the implementation. By calling my smoothing function in MATLAB with the calculated curvature data from OIP, MATLAB had the same result.

Now with the limit of range in commit fcd1513, the charts are the following:
![5 afterFix curvature](https://user-images.githubusercontent.com/64251870/95001860-6c1a5900-05ce-11eb-8083-ccc70295afb1.png)
![6 afterFix curvatureChange](https://user-images.githubusercontent.com/64251870/95001861-6e7cb300-05ce-11eb-9bce-49bbd9b43fe5.png)
The smoothing of perfect data works better.

As well, I have an example of a sketched alignment:
![7 sketch curve](https://user-images.githubusercontent.com/64251870/95002236-f617f100-05d1-11eb-82dd-7502fb6e32e9.png)
![8 sketch curvature](https://user-images.githubusercontent.com/64251870/95002241-fb753b80-05d1-11eb-8032-bc25d7a28a9a.png)
![9 sketch curvatureChange](https://user-images.githubusercontent.com/64251870/95002242-fca66880-05d1-11eb-9651-9ec83261501e.png)
The quality of smoothing is okay, especially in the range of curve length 0 to 400 short disturbances in curvature are smoothed well.

christophKaiser committed:
Introduced class SplineInterpretationElement
The objects of this class hold the information of each element, which is identified by SplineInterpretation::convertSketchToAlignment(). The class will be used in the next commit.

christophKaiser committed:
Added functions to identify the element endpoints from the curvature change.
doxygen comments in the header file will be added later.

pjanck commented:
This looks very nice. The graphs of the "real world" example seem very wobbly, I'm not sure if you will even recognize the straight section in the middle. However, the synthetic results look very good. 

A small hint regarding visualization: it is suboptimal to use grey and black - when copying this information can get hazy. It is better to use different line patterns, like full line, dashed line, dotted line, dash-dot etc.

pjanck commented on Rendering/Tools/SplineInterpretation.cpp:
```diff
@@ -345,6 +370,185 @@ std::vector<double> OpenInfraPlatform::UserInterface::SplineInterpretation::nume
+}
+
+std::vector<SplineInterpretationElement> OpenInfraPlatform::UserInterface::SplineInterpretation::obtainElementsFromIndicator(
+	std::vector<int> indicator,
+	std::vector<std::pair<double, double>> lengthsWithCurvatures) const throw(...)
```
if you are not changing a variable within the function, it is advisable to pass it as `const &` - saves the copying time for the processor.

```suggestion
	const std::vector<int>& indicator,
	const std::vector<std::pair<double, double>>& lengthsWithCurvatures) const throw(...)
```

Primitive types (like `int` and `double`) don't need the reference `&`.

christophKaiser replied:
Done commit 154ca21

pjanck commented on Rendering/Tools/SplineInterpretationElement.cpp:
```diff
@@ -0,0 +1,75 @@
+
+
+// PUBLIC FUNCTIONS WITH ACCESS TO VARIABLES
+void OpenInfraPlatform::UserInterface::SplineInterpretationElement::setLength(double newLength) throw(...){
+	length = newLength; }
```
Such a function does not throw, there is no `throw` statement within it. Use `throw()` to denote that.

christophKaiser replied:
In commit b492fab, I removed the `throw` from the get- and set-functions.
Do you mean with "`throw` statement" something like the following lines?
```
if(badCase)
   throw oip::UnhandledException(myVariable);
```

pjanck replied:
Yes, sth like this. Beware, though: if you use `throw()` as an indicator to a function definition, then neither the function itself or the functions it calls are not allowed to throw. Which means - if you are not throwing anything in the function and you do not call another function, where you are unsure if it throws, you can use the `throw()` indicator.

pjanck reviewed (commented):
Small suggestions for code improvement.

christophKaiser committed:
Manipulate first and last point of curvature.
As shown in the charts from 4. Oct. [1], the first and last value of the curvature in CAD data makes a huge jump, which has a large impact after smoothing. By overwriting the end values with their neighbors, this effect doesn't appear. In cases where a different value is correct, the mistake by overwriting is small.
At the moment, I don't know where the jumping end values come from - maybe it's an issue how the B-Spline is used and how the derivatives and curvature are defined.
[1] https://github.com/tumcms/Open-Infra-Platform/pull/202#issuecomment-703169881

christophKaiser committed:
Renamed function averageOfCurvautreChange to indicateDataByAverage
Generalized the function name and its variables for further use in the analysis of curvature.

christophKaiser committed:
Added function obtainThreshold
Moved the calculation of the threshold (previous named as curvatureZero) into a own function for later use.

christophKaiser committed:
Removed second approximation
Originally, in the function call of 'computeCurvatureOfBSplineCurveWithKnots' the points should be 'bsplinePoints' (result of first B-Spline calculation). As mistake, there was the 'controlPoints' (sketch data) used, thus the curvature was from the first approximation B-Spline instead the second approximation.
However, now with moving average the first approximation provides better results.

christophKaiser committed:
Modified indicateDataByAverage
In the new version, the function can identify about 0, less than 0 and greater than 0.
The indicator in correctShotElement is take into abs(), because this indicator does only hold about 0 or not (=1).

christophKaiser committed:
Added variables to class SplineInterpretationElement ...
... with matching get- & set-functions.
As well, renamed exiting variables and functions according to coding guideline (post-pended underscore).

christophKaiser committed:
Added function identifyElemntTypes with subfunctions
Parameters of straights and arcs will be calculated, clothoids are missing (but coming soon)

christophKaiser committed:
Fixed typing error

christophKaiser committed:
Added functions of obtainClothoid
As well, SplineInterpretationElement gets variables (including get- / set- functions) to store information of clothoid.
At this stage, the interpretation algorithm is complete (according to current design plan).

christophKaiser committed:
Added function printElementsInConsoleWindow
Prints all obtained information as result in the console window.

christophKaiser committed:
Turned off the debug functions

christophKaiser committed:
Fixed error in calculation of arc center
During theoretical preparing, I didn't take attention that mid points P1 and P2 have different distances to the center.

christophKaiser committed:
Introduced 'angle' as parameter of arc element
At least, this information is very helpful for drawing the arc in AutoCAD.

christophKaiser commented:
Now, there are results from the data of the charts which I showed you a few days ago ([link to comment](https://github.com/tumcms/Open-Infra-Platform/pull/202#issuecomment-703169881)).

First the results to the calculated data:
![01 KreisKlothoideGerade Berechnet Ergebnis](https://user-images.githubusercontent.com/64251870/95626109-edb03200-0a79-11eb-95dc-33e63f2e27fc.png)

There is a drawing (with AutoCAD) which shows the resulting elements from the calculated data in orange. The element endpoints are highlighted with a circle point. The ends of the clothoids don't match perfect, maybe it's an issue of accuracy in my Excel calculation of the clothoid curve points (support tool to draw clothoids in AutoCAD).
Under the orange line is a white line which shows the original alignment.
![01 KreisKlothoideGerade Berechnet Ergebnis Zeich](https://user-images.githubusercontent.com/64251870/95626360-66af8980-0a7a-11eb-92e5-1006c9caa4dd.png)

Here are the results from the sketched data:
![01 KreisKlothoideGerade Skizze Ergebnis](https://user-images.githubusercontent.com/64251870/95626120-f1dc4f80-0a79-11eb-9323-96718d29e565.png)

christophKaiser committed:
Removed unused variable endPoint from obtainClothoid

christophKaiser committed:
Added test sketches which can be loaded to interpret

christophKaiser commented:
The last commit 917a313 is the source code at end of the BA.

christophKaiser commented:
The following listing is an overview of pending work for optimisation, which was collected in the above comments and commit messages. The order of my tasks represents more or less the priority according to my opinion. However, it’s free for discussion.
- [ready for review, commits 6bab0e3 and a93fbd7] The class `SplineInterpretation` doesn't appear in doxygen (help wanted)
- [ready for review, commit 2feabf6] Complete doxygen comments in `SplineInterpretation.h`.
- [ready for review, commit 07889a5] In `SplineUtilities`: Further investigation to include all really needed files / libraries might be reasonable. (in more detail: commit f13a07d)
- [canceld, get wrong information] Modify function definitions in cpp-files by use of `namespace` or `using namespace`. This should short the definitions for more readability. (Experience from current CiE 1 course.)
- [ready for review, commit 51a8d46] Add exception handling if loading of a file fails. (`SplineInterpretation::loadControlPointsFromFile()`, line 208)
- [ready for review, commit ccc183c] Add exception handling if `controlPoints` is an empty vector after loading. (`SplineInterpretation::convertSketchToAlignment()`, line 44)
- [ready for review, commit 154ca21] Rework `const` and `throw`. (regarding to pjanck's comment [above](https://github.com/tumcms/Open-Infra-Platform/pull/202#pullrequestreview-501728744))
- The final result (input points and alignment elements) should be displayed on the viewport.

Optional (currently not planed)
- Performance work: The function `convertSketchToAlignment()` needs about 2 minutes for the result in my case and with 168 control points (example `001 Testsketch ArcClothoidStraight calculated.sketch`). From this time, the function `splineUtilities.computeCurvatureOfBSplineCurveWithKnots(…)` needs about 1:40 minutes. At this point, further observation and analysis of the code is necessary. Maybe, the calculation of the derivatives and their basis functions must be redesigned.
- Create a unit test, if displaying on the viewport was implemented successful.
- Implementation of sketch input.

EDIT 7 Jan 2021: Added progress in square brackets.
EDIT 8 Jan 2021: 'Performance work' was moved to optional (not planed) (according to today's meeting)
EDIT 26 Feb 2021: Updated progress in square brackets.

christophKaiser committed:
Added missing doxygen comments.

christophKaiser committed:
Refactoring of include statements in SplineUtilities
Until now, all includes came from higher calling instances.
Now, the class has all needed include statements in its own definition. "pragma once" prevents multiple inserts.

christophKaiser committed:
Added missing "pragma once" in SplineInterpretation.cpp

christophKaiser committed:
Added comments in SplineInterpretation::loadControlPointsFromFile, ...
removed typos & improved formatting of comments.

christophKaiser committed:
Added exception handling to loadControlPointsFromFile.
The function returns an empty vector especially in the cases:
- Load dialog was canceled
- File wasn't able to open
- Syntax in file doesn't fit

christophKaiser committed:
Added exception handling to obtainControlPoints ...
respectively to its call: If an empty vector is the return variable (sign for error), the main function will be canceled.

christophKaiser committed:
Added a message box to inform about finished calculation.

christophKaiser commented:
Updated the comment with pending work from 23 Nov 2020 (see [here](https://github.com/tumcms/Open-Infra-Platform/pull/202#issuecomment-732456798)).
Additional commits from above:
- Small correction and comment optimation in commits 97447bf and 3ae2877.
- Additional message box to inform about finished calculation in commit 9988bca.

pjanck commented on Rendering/Tools/SplineInterpretation.cpp:
```diff
@@ -15,6 +15,8 @@
 */
 
+#pragma once
+
```
No need for this - this is useful in headers.

christophKaiser replied:
Reverted in commit 00bf496.

pjanck commented on Rendering/Tools/SplineInterpretation.h:
```diff
@@ -121,6 +124,14 @@ namespace OpenInfraPlatform
+			 *
+			 * \return		A vector of \c double wiht the second values of the pair vector.
+			 */
 			std::vector<double> obtainCurvatureFrom_lengthWithCurvatures(const std::vector<std::pair<double, double>>& lengthsWithCurvatures) const throw(...);
```
This is easily obtained with a `std::transform` locally.

christophKaiser replied:
The function calls are replaced to `std::tansform` in commit 75dcd42. As well, one function overload of obtainThreshold is removed entirely, instead `std::transform` is used at its previous call.

pjanck reviewed (commented):
Thank you for the doxy-comments - now it is much clearer, what the individual functions do and produce.

christophKaiser committed:
Revert "Added missing 'pragma once' in SplineInterpretation.cpp"

This reverts commit 97447bf3081fbc58423b707ef2cad771a543dcf6 according to pjanck's comment.

christophKaiser committed:
Replaced the function obtainCurvatureFrom_lengthWithCurvatures ...
by local std::transform; according to pjanck's comment.
As well, one overload of obtainThreshold was removed, now its task is done by a std::transform at the place of the function call.

christophKaiser committed:
Merge branch 'development' into addSplineCurvature

christophKaiser commented:
**About 'The class `SplineInterpretation` doesn't appear in doxygen'** (see pending work [here](https://github.com/tumcms/Open-Infra-Platform/pull/202#issuecomment-732456798))

The source files of the classes `SplineInterpretation` and `SplineInterpetationElement` are stored in the directory `/Rendering/Tools/`.

With the following line in the general Doxygen settings all content in folders 'Tools' is excluded (besides other folders), what is the reason why the classes doesn't appear in the generated documentation.
https://github.com/tumcms/Open-Infra-Platform/blob/90e74889a85fb7d1044419a0297c657c95d38355/cmake/ConfigureDoxygen.cmake#L76

If I remove the item `*/Tools/*` from this list (immediately before the last item), the mentioned classes appear in the documentation. However, I'm not sure whether this is the best solutuion because at somepoint in past someone (probably) had a reason to exclude that folders (there are other Tools folders in other subdirectories of the project).

An alternative solution might be to move these source files into another folder which isn't excluded from the documentation - or just rename the appropriate folder to fall out of the exclude pattern.

Probably, there might be another solution as well which I don't know. @pjanck  At this point, I want to ask for help to get / found the best solution for that issue - especially in context and with respect to the hole project.

pjanck commented:
I propose to move your content to `Core/SplineInterpretation` directory.

Be sure to amend `Core/CMakeLists.txt` to include the files in the solution when generating with CMake.

christophKaiser committed:
Moved source files of SplineInterpetation ...
from directory /Rendering/Tools/ into /Core/src/SplineInterpretation/.
Included the new folder in the CMakeLists as part of the sub-project Core.

christophKaiser committed:
Adjusted the namespace of SplineInterpretation ...
according to its new directory path.

pjanck commented on Core/CMakeLists.txt:
```diff
@@ -28,13 +28,15 @@ file(GLOB Core_DataManagement_General_Source 		src/DataManagement/General/*.*)
 file(GLOB Core_IfcGeometryConverter_Source			src/IfcGeometryConverter/*.*)
 file(GLOB Core_OffConverter_Source					src/OffConverter/*.*)
 file(GLOB Core_Exception_Source						src/Exception/*.*)
+file(GLOB Core_SplineInterpretation_Scource			src/SplineInterpretation/*.*)
```
Typo - a "c" too many :)

christophKaiser replied:
Done in commit 347f2ab

pjanck commented on Core/src/SplineInterpretation/SplineInterpretation.cpp:
```diff
@@ -454,20 +444,14 @@ std::tuple<std::vector<int>, double> OpenInfraPlatform::UserInterface::SplineInt
-}
-
-double OpenInfraPlatform::UserInterface::SplineInterpretation::obtainThreshold(std::vector<double> data) const throw(...) {
+double OpenInfraPlatform::Core::SplineInterpretation::SplineInterpretation::obtainThreshold(std::vector<double> data) const throw(...) {
```
Beginning with vs2019 (which we'll be switching to soon) the `throw(...)` should be replaced with `noexcept(false)`. Please replace in all new files written by you accordingly.

christophKaiser replied:
Done in commits 94afcd3 and 615d2b1

pjanck reviewed (changes requested):
Some small suggestions.

christophKaiser committed:
Fixed typo
Regarding to commit 'Moved source files of SplineInterpretation'

christophKaiser committed:
Reworked throw in SplineInterpretationElements

christophKaiser committed:
Reworked throw in SplineInterpretation

christophKaiser committed:
Reworked const and & of function parameters

christophKaiser committed:
Fixed 5 warnings in the VS error-list, ...
which showed up after the project was built.

pjanck commented on Core/src/SplineInterpretation/SplineInterpretation.cpp:
```diff
@@ -263,15 +263,15 @@ std::vector<carve::geom::vector<3>> OpenInfraPlatform::Core::SplineInterpretatio
 		// in case of an error, return an empty vector
 		std::vector<carve::geom::vector<3>> pointsEmpty;
 		return pointsEmpty;
 	}
```
It is advisable to at least log that an error occured. Use sth similiar to:

    BLUE_LOG(warning) << "error occured ... ";

Be as specific as possible, so that the future programmer knows where to look.

christophKaiser replied:
Done in commit 643376a.

pjanck commented on Core/src/SplineInterpretation/SplineInterpretation.cpp:
```diff
@@ -368,10 +368,11 @@ size_t OpenInfraPlatform::Core::SplineInterpretation::SplineInterpretation::vari
 		gamma[h - 1] /= (2 * N);
```
```suggestion
		gamma[h - 1] /= (2. * N);
```

It is advisable to use floating numbers when dividing, so that the result is not rounded.

christophKaiser replied:
Done in commit e4fc298.

pjanck commented on Core/src/SplineInterpretation/SplineInterpretation.cpp:
```diff
@@ -771,16 +768,19 @@ SplineInterpretationElement OpenInfraPlatform::Core::SplineInterpretation::Splin
+	//   0 is just a preset value;
+	//   the actual one will be obtained surely in one of the following if-cases,
+	//   because of the requirement of one arc at one clothoid end
+	double radius = 0;
```
```suggestion
	double radius = 0.;
```

Initialize floating variables with floating numbers. Always.

christophKaiser replied:
Done in commit e4fc298.

pjanck commented on Core/src/SplineInterpretation/SplineInterpretation.cpp:
```diff
@@ -862,12 +862,12 @@ carve::geom::vector<3>  OpenInfraPlatform::Core::SplineInterpretation::SplineInt
-	double direction) const throw(...)
+	const double direction) const noexcept(true)
 {
 	return carve::geom::VECTOR(cos(direction*M_PI / 180), sin(direction*M_PI / 180), 0.0);
```
```suggestion
	return carve::geom::VECTOR(cos(direction*M_PI / 180.), sin(direction*M_PI / 180.), 0.0);
```

How come that direction is in degrees? Naturally, in C, angles are in radians. Where does this come from?

christophKaiser replied:
Done in commit 3ca85d7.
These angles are calculated in the function `angleOfVectors2D` where they were converted to degree manually.
During the development process, I was focused on testing with CAD drawings where usually degree is used. According to https://standards.buildingsmart.org, both degree and radiant is possible in an ifc file. At this point, I probably missed to doublecheck what is used in the ifc example files.
With the current changes, everything program internal works in radian and just the print output to the console window is in radian and degree (just in case that the value in degree might still be usefull in some further usage or human interpretation).

pjanck replied:
Right, I understand. Rule of thumb:

- Internally, OIP works in meters and radians. 
- Whenever we print, we may decide and reformat to any other unit.

pjanck reviewed (changes requested):
Small suggestions. Should be applied everywhere ( I'm not sure I managed to get all occurrences).

christophKaiser committed:
Added a log message in case of loadFromFile fails

christophKaiser committed:
Adjusted numbers to floating numbers in case of double calculation.

christophKaiser committed:
Adjusted unit of directional angle to radian.
Now, all calculation and information storage works in radian; just the print output to the console window has radian and degree. To get the angle in degree two functions were added to SplineInterpretationElement.

pjanck commented on Core/src/SplineInterpretation/SplineInterpretationElement.cpp:
```diff
@@ -107,6 +107,9 @@ void OpenInfraPlatform::Core::SplineInterpretation::SplineInterpretationElement:
 double OpenInfraPlatform::Core::SplineInterpretation::SplineInterpretationElement::getDirection() const noexcept(true) {
 	return direction_; }
 
+double OpenInfraPlatform::Core::SplineInterpretation::SplineInterpretationElement::getDircetionDegree() const noexcept(true) {
```
typo in function name

christophKaiser replied:
Function was removed by changes according to the suggestion of using default parameter (see next below).
Done in commit d8bb82e.

pjanck commented on Core/src/SplineInterpretation/SplineInterpretationElement.cpp:
```diff
@@ -128,6 +131,9 @@ void OpenInfraPlatform::Core::SplineInterpretation::SplineInterpretationElement:
 
+double OpenInfraPlatform::Core::SplineInterpretation::SplineInterpretationElement::getAngleDegree() const noexcept(true) {
+	return angle_ * 180.0 / M_PI; }
+
```
```suggestion
double OpenInfraPlatform::Core::SplineInterpretation::SplineInterpretationElement::getAngle( bool inDegrees = false ) const noexcept(true) {
	return angle_ * (inDegrees ? 180.0 / M_PI : 1.) ; }
```

Function with a default parameter could be considered. (syntax may be wrong, but I believe you get the idea)

christophKaiser replied:
Thanks for that suggestion! The concept of default parameter was new to me.
Done in commit d8bb82e.

pjanck commented on Core/src/SplineInterpretation/SplineInterpretationElement.cpp:
```diff
@@ -128,6 +131,9 @@ void OpenInfraPlatform::Core::SplineInterpretation::SplineInterpretationElement:
+	return angle_ * 180.0 / M_PI; }
+
 
 void OpenInfraPlatform::Core::SplineInterpretation::SplineInterpretationElement::setIsCCW(const int isCCW) noexcept(true) {
```
how come `isCCW` is an `int` and not a `bool`?

christophKaiser replied:
To be honest, I can't remember.
It's changed to `bool` in commit 1d19e35.

christophKaiser committed:
Reworked getDirection() and getAngle() ...
by applying the concepts of Default Argument, and Conditional (or Ternary) Operator (or colloquial 'Immediate-if').

christophKaiser committed:
Fixed a few syntax mistakes in Doxygen comments.

christophKaiser committed:
Changed type of isCCW from int to bool.
(isCCW = is counter clockwise)

christophKaiser commented:
Additionally, to the comments in the conversations, commit f5f402c fixes a few syntax mistakes in the Doxygen comments. Probably I didn't recognise them before because they are in private functions which don't appear in the documentation by default.

pjanck reviewed (approved):
I think that this is now done.

pjanck commented on Core/src/IfcGeometryConverter/SplineConverter.h:
```diff
@@ -565,6 +364,8 @@ namespace OpenInfraPlatform
 
 					std::shared_ptr<PlacementConverterT<IfcEntityTypesT>> placementConverter;
 
+					SplineUtilities splineUtilities;
+
```
I'd propose to make the functions `static` and omit the need for a member here.

christophKaiser replied:
Done in commit 4f9851b.
Now, all functions of SplineUtilities are static because they don't store information in member variables of SplineUtilities. As well, these functions don't call other non-static functions, thus your suggestion is comprehensible - thanks.

pjanck commented on Core/src/IfcGeometryConverter/SplineUtilities.h:
```diff
@@ -0,0 +1,768 @@
+				 * \return		The array of curve points, which can be rendered in a viewport after correction by \c GeomUtils::appendPointsToCurve.
+				 */
+				 // B-Spline curve definition according to: http://mathworld.wolfram.com/B-Spline.html
+				std::vector<carve::geom::vector<3>> computeBSplineCurveWithKnots(
+					const int& order,
+					const std::vector<double>& knotArray,
+					const std::vector<carve::geom::vector<3>>& controlPoints,
+					const uint32_t& numCurvePoints,
+					const double& accuracy) const throw(...)
```
For example, this can definitely be a `static` function.

christophKaiser replied:
Done in commit 4f9851b.

pjanck reviewed (approved):
I say we merge this and move forward with implementing B-Spline curves from IFC (after addressing the proposals below).

christophKaiser committed:
Modified functions in SplineUtilities to static functions.
Adjusted all calls of these functions.

christophKaiser committed:
Merge remote-tracking branch 'tumcms/development' into addSplineCurvature

christophKaiser committed:
Change to 2021
Adapted changes of pull request #386 to local files which aren't merged into tumcms yet.

christophKaiser committed:
Added files of unit test
Added a .sketch file to test the program, the expected result is staved in the .txt file.
A short description how to do the test is included in the .md file, because a implementation in a source file is still pending.

christophKaiser committed:
Added an adaption of the console size before printing of the results
While writing the description about the test (md-file), I recognized that the manual change of the Screen Buffer Size could be annoying to other users.
The added code changes autonomous the Screen Buffer Height dynamic depending on the number of alignment elements.
(Screen Buffer Height is the number of stored lines in the console. A scroll bar appears when the number of lines doesn't fit into the current window size.)

christophKaiser commented:
Commit cb3a5a4 addes files to do a [manual] unit test of the function 'Core::SplineInterpretation::SplineInterpretation::convertSketchToAlignment' including its subfunctions. Commit ebfac5a provides a optimization of the print output in the console window.
Issues about the pending work (which aren't urgent) are created (#408 #409 #410 #411 )- according to the last dev meeting.

The current commit of the branch `tumcms/development` was merged into this branch with commit c4da9bd.
From my side, this pull request is ready to be merged into the devolpment branch.

christophKaiser commented:
I forgot to mention: The provided example files of Stroke to Alignment will become available after build of the project */Copy/CopyOpenInfraPlatformUIResources*:
![BuildProject](https://user-images.githubusercontent.com/64251870/111284133-23b49c00-8640-11eb-952a-8ae80f9e0428.png)

Once built, you can start the function via the menu *Tools* > *Convert Stroke to Alignment*. After chooseing of the option *Load*, you should get the following open dialog which shows you three files in the build directory:
![FilesInOpendialog](https://user-images.githubusercontent.com/64251870/111284745-b2c1b400-8640-11eb-93bf-c5bf10884a93.png)

You can store additional sample files in this folder, without effects on the directory which is tracked by git.
